Add Op Diff

CelysPr opened the pull request:
### PR Category
Operator

### Type of Change
Add new operator

### Description
Implement diff(n) by **recursively calling diff(n - 1).**
Performance is suboptimal and slow. An alternative implementation using **convolution** resulted in further performance degradation (speedup < 0.1) and significant accumulation error.

### Issue
#393 

### Progress

- [x]  Performance optimization

### Performance
Operator is accurate for larger values of n (tested for n = 0 to 50).
<img width="991" alt="image" src="https://github.com/user-attachments/assets/f832ad84-7380-407b-a40d-71944bcb564a" />
Performance on A800 (cloud GPU):
<img width="844" alt="image" src="https://github.com/user-attachments/assets/c34804a2-72b5-4563-8fbc-631834a6dd3a" />
<img width="844" alt="image" src="https://github.com/user-attachments/assets/c79e1259-dc37-4df1-bf47-a2b95981ba0f" />
<img width="835" alt="image" src="https://github.com/user-attachments/assets/53d281be-8eae-45cc-ad04-02aa5b5c8964" />
<img width="833" alt="image" src="https://github.com/user-attachments/assets/b7676d1e-3ffd-4bd2-806a-4e81db873491" />
<img width="857" alt="image" src="https://github.com/user-attachments/assets/de95d67d-83d4-48b8-8816-f42b7afb0460" />
<img width="850" alt="image" src="https://github.com/user-attachments/assets/9dbeeec6-2f26-4bc6-9ce9-c33533299533" />

StrongSpoon commented:
I'm not sure if torch.diff belongs to reduction operators. is it possible to implement it as a pointwise one?

0x45f commented:
@CelysPr Have you tried to fix the performance issues recently?

CelysPr commented:
> Hi, please let us know if you are still working on this contribution. Thanks.

I will continue to work on it this week. Sorry for my late reply

w1120029931-bit reviewed (changes requested):
The code architecture has been adjusted,
test_reduction_ops.py
does not exist currently, Each operator requires separate functional test files and performance test files. Please refer to the relevant code modifications in the repository

CelysPr committed:
[Operator] Replace diff with tuned 2D kernel + edge-case fixes

Replace the auto-generated diff implementation with a hand-written, libtuner-tuned
variant ported from the legacy dev_diff branch and modernized to current FlagGems
conventions.

Kernel design:
- diff_kernel_1d / diff_kernel_2d operate on the inner dimension after
  dim_compress() rotates the target axis last.
- 2D grid (rows x col-tiles) exposes both M and N parallelism.
- Autotuned via libtuner; tune entries added for _nvidia/_metax/_iluvatar.
- Multi-step n>1 reuses a single full-width output buffer via copy_().

Edge-case bug fixes (now matches torch.diff semantics, verified empirically):
- n < 0 raises 'order must be non-negative but got X'
- 0-d input raises 'diff expects input to be at least one-dimensional'
- n == 0 short-circuits BEFORE prepend/append, returning inp.clone()
- n >= shape[dim] returns a zero-length tensor instead of raising

Tests (tests/test_diff.py):
- INT dtype coverage added (torch.int16, torch.int32)
- prepend/append coverage retained
- explicit edge-case tests added

Operator metadata:
- conf/operators.yaml: drop KernelGen label since implementation is now
  hand-written + tuned (stage beta 5.1 retained)

CelysPr committed:
Refined Operator diff & Test/Bench Cases

CelysPr committed:
Merge branch 'flagos-ai:master' into dev_diff

w1120029931-bit committed:
Merge branch 'master' into dev_diff

w1120029931-bit committed:
Merge branch 'master' into dev_diff

w1120029931-bit commented on src/flag_gems/ops/diff.py:
```diff
@@ -1,234 +1,140 @@
 import logging
```
这里缺少 KernelGen 文件头。若该文件由 KernelGen 生成，请在第一行添加 `# Generated by KernelGen`。

<sub>Rule: `KERNEL_HEADER`</sub>

CelysPr replied:
diff.py不是kernelgen生成的，未添加注释。

w1120029931-bit commented on benchmark/test_diff.py:
```diff
@@ -24,3 +33,14 @@ def test_diff():
+
+
+@pytest.mark.diff
+def test_diff_n2():
```
这里的 `op_name="diff_n2"` 与 `operators.yaml` 中的算子 id `diff` 不一致。当前 YAML 里没有 `diff_n2` 这个 id，建议改为 `op_name="diff"`，保持 benchmark 的 op_name 与 yaml id 一致。

<sub>Rule: `BENCH_OP_NAME`</sub>

CelysPr replied:
已修改

w1120029931-bit reviewed (commented):
## FlagGems 自动 Review

**Operator:** `diff`
**结论:** COMMENT

### Summary
- Errors: 2
- Warnings: 0
- Suggestions: 0

CelysPr committed:
[Operator] Diff: Modify op_name in test_diff.py

CelysPr committed:
Merge branch 'master' into dev_diff

CelysPr committed:
Merge branch 'master' into dev_diff

Signed-off-by: CelysPr <128201073+CelysPr@users.noreply.github.com>

w1120029931-bit committed:
Merge branch 'master' into dev_diff
